---
title: "Setup"
author: "Carolina Lobato"
date: "2023-07-20"
output: html_document
---

# Load Libraries
```{r loadlib, include = F}
library(phyloseq)
library(microbiome)
library(microbiomeutilities)
library(tidyverse)
library(ggvenn)
library(RColorBrewer)
```

# Pacbio

## Create Phyloseq
```{r}
load("outputs/r/1_dada2/phy-obj/metadata.RData") # df
load("outputs/r/1_dada2/phy-obj/feature-table-corr.RData") # st.nochim.corr
load("outputs/r/1_dada2/phy-obj/tax-table.RData") # tax

asv_tab <- t(st.nochim.corr)
rownames(asv_tab) <- paste(
  "ASV ", 1:length(rownames(asv_tab)),
  ": ",
  rownames(asv_tab),
  sep = "")
rownames(asv_tab)[1:5]
rownames(asv_tab) <- sub(": [CAGT]+", "", rownames(asv_tab))
rownames(asv_tab)[1:5]

tax_tab <- as.data.frame(tax)
rownames(tax_tab) <- paste(
  "ASV ", 1:length(rownames(tax_tab)),
  ": ",
  rownames(tax_tab),
  sep = "")
rownames(tax_tab)[1:5]
rownames(tax_tab) <- sub(": [CAGT]+", "", rownames(tax_tab))
rownames(tax_tab)[1:5]
tax_tab <- as.matrix(tax_tab)

sq <- Biostrings::readDNAStringSet(
  "outputs/r/1_dada2/phy-obj/dna-seq2.fasta",
  format = "fasta",
  nrec = -1L,
  skip = 0L,
  seek.first.rec = F,
  use.names = T,
  with.qualities = F)

ps <- phyloseq(otu_table(asv_tab, 
                         taxa_are_rows = T),
               tax_table(tax_tab),
               sample_data(df),
               sq)

ps # 445 taxa and 136 samples
# save(ps, file = "outputs/r/2_setup/ps.RData")

## Read stats ps
ss <- sort(sample_sums(ps))
ss <- as.matrix(ss)
colSums(ss) # total reads = 204746
summary(ss) # mean reads/sample = 1505.5; min = 4; max = 7315
```

## Check Contaminants
```{r}
load("outputs/r/2_setup/ps.RData") 

ps_new <- subset_samples(ps, tissue == "") 
ps_new <- prune_taxa(taxa_sums(ps_new) > 0, ps_new);ps_new # 25 taxa and 4 samples
check <- ps_new@sam_data[["genotype"]];check # n = 4 

contam.list <- c()
metadata <- sample_data(ps_new)
categories <- unique(na.omit(metadata[["genotype"]]))

for (item in categories)
{
  print(paste0("Identifying Taxa for ", item))
  ps.subset <- subset_samples(ps_new, genotype == item)
  subset.list <- core_members(ps.subset, detection = 0 , prevalence = 0)
  print(paste0("No. of Taxa in ", item, ": ", length(subset.list)))
  contam.list[[item]] <- subset.list
}
print(contam.list)

common.asv <- Reduce(intersect, contam.list); common.asv # n = 0
all.asv <- Reduce(union, contam.list); all.asv # n = 25

genotypes <- c("M")
other_genotypes <- setdiff(names(contam.list), genotypes)
set1 <- Reduce(intersect, contam.list[genotypes])
set2 <- unique(unlist(contam.list[other_genotypes]))
setdiff(set1, set2)

# write.table(contam.list$M,
#             sep = ";",
#             "outputs/r/2_setup/mock-list.csv") # 23 taxa
# 
# write.table(contam.list$Ctl,
#             sep = ";",
#             "outputs/r/2_setup/ctl-list.csv") # 2 taxa


## Plot
mycolors <- colorRampPalette(brewer.pal(9, "Paired"))(9) # 9 Genera
ps_rel <- microbiome::transform(ps_new, "compositional") # 25 ASVs

FigS2 <- plot_bar(ps_rel, 
         x = "sample_name", 
         fill = "Genus") + 
  labs(x = "", color = "") +
  scale_color_manual(values = mycolors,
                     aesthetics = "fill",
                     na.translate = T) +
  theme_classic() +
  theme(
    axis.text.x = element_text(size = 12, angle = 45, hjust = 1),
    legend.position = "right",
    legend.text = element_text(size = 12),
    legend.title = element_text(size = 12, face = "bold"),
    strip.background = element_blank()) +
  guides(fill = guide_legend(ncol = 1)); FigS2

# ggsave(
#   "outputs/r/SI/figS2.png",
#   width = 45,
#   height = 31,
#   units = "cm",
#   scale = 1,
#   dpi = 600)

# Contaminants NC:
# ASV 94: Sphingomonas
# ASV 111: Sphingomonas

# Mock:
# Pseudomonas aeruginosa ASV 355
# Escherichia coli ASV 329, ASV 408, ASV 420, ASV 428, ASV 435
# Salmonella enterica ASV 258, ASV 430
# Lactobacillus fermentum ASV 310, ASV 409, ASV 412, ASV 422, ASV 424
# Enterococcus faecalis ASV 186
# Staphylococcus aureus ASV 192, ASV 326, ASV 334
# Listeria monocytogenes ASV 227, ASV 282, ASV 385
# Bacillus subtilis: ASV 171, ASV 280, ASV 289
```

## Subset Controls, Assign New ASV Names and Clean
```{r}
load("outputs/r/2_setup/ps.RData") 

ps_sub <- subset_samples(ps, tissue != "")
ps_sub <- prune_samples(sample_sums(ps_sub) > 0, ps_sub)
ps_sub <- prune_taxa(taxa_sums(ps_sub) > 0, ps_sub)
ps_sub # 422 taxa and 132 samples

## Read Stats ps_sub
sb <- sort(sample_sums(ps_sub))
sb <- as.matrix(sb)
colSums(sb) # total reads = 203031
summary(sb) # mean reads/sample = 1538.1

bh <- microbiomeutilities::format_to_besthit(ps_sub)
taxa_names(bh) <- sub(":", ": ", taxa_names(bh))
taxa_names(bh)[1:5]

sample_data(bh)$total_reads <- sample_sums(bh)
metadata_bh <- sample_data(bh)

bh_pb <- subset_taxa(bh, rownames(tax_table(bh)) != c("ASV 94: Sphingomonas"))
bh_pb <- prune_taxa(taxa_sums(bh_pb) > 0, bh_pb)
bh_pb # 421 taxa and 132 samples

bh_pb <- subset_taxa(bh_pb, rownames(tax_table(bh_pb)) != c("ASV 111: Sphingomonas"))
bh_pb <- prune_taxa(taxa_sums(bh_pb) > 0, bh_pb)
bh_pb # 420 taxa and 132 samples

bh_pb <- subset_taxa(bh_pb, Domain != "Eukaryota")
bh_pb <- prune_taxa(taxa_sums(bh_pb) > 0, bh_pb)
bh_pb # 419 taxa and 132 samples

tax_table(bh_pb)

## Read Stats bh_pb
sh <- sort(sample_sums(bh_pb))
sh <- as.matrix(sh)
colSums(sh) # total reads = 202111
summary(sh) # mean reads/sample = 1531.1; min = 23; max = 7283
sample_data(bh_pb)$total_reads <- sample_sums(bh_pb)
metadata_bh_pb <- sample_data(bh_pb)

final.reads <- sum(metadata_bh_pb$total_reads)

# save(bh_pb, file = "outputs/r/2_setup/bh-pb.RData")

# refseq(bh_pb) %>%
#   Biostrings::writeXStringSet("outputs/r/2_setup/bh-pb-seq.fasta",
#                               append = F,
#                               compress = F,
#                               compression_level = NA,
#                               format = "fasta")

# tax_table(bh_pb) %>% as.data.frame() %>%
#   write_csv(file = "outputs/r/2_setup/bh-tax-train-set.csv")
```

## Check Samples
```{r}
load("outputs/r/2_setup/bh-pb.RData"); bh_pb # 419 taxa and 132 samples
load("outputs/r/2_setup/bh-amp.RData"); bh # 5297 taxa and 451 samples

# Check samples per Genotype
sg <- as.data.frame(bh_pb@sam_data[["genotype"]]) #%>% summarise(count = n())
names(sg)[1] ="Genotype"
n_sg <- as.data.frame(table(sg$Genotype))

# Check sample overlaps
Community <- as.data.frame(unique(bh@sam_data[["genotype"]]))
colnames(Community) <- c("Gen") 

Culturing <- as.data.frame(unique(bh_pb@sam_data[["genotype"]]))
colnames(Culturing) <- c("Gen") 

sets <- list(Community = sort(as.character(unique(bh@sam_data[["genotype"]]))),
             Culture =  sort(as.character(unique(bh_pb@sam_data[["genotype"]]))))

FigS1 <- ggvenn(sets,
                 fill_color = c("#0073C2FF", "#CD534CFF"),
                 fill_alpha = 0.7, 
                 show_percentage = F,
                 stroke_size = 0.5, set_name_size = 5, text_size = 4); FigS1

# ggsave(
#   "outputs/r/SI/figS1.png",
#   width = 15,
#   height = 15,
#   units = "cm",
#   scale = 1,
#   dpi = 600)
```

# Assignment

## Import Data
```{r}
load("outputs/r/2_setup/bh-amp.RData"); bh # 5297 taxa and 451 samples
load("outputs/r/2_setup/bh-pb.RData"); bh_pb # 419 taxa and 132 samples

# Get taxonomic information
taxo.amp <- tax_table(bh) %>% 
  as.data.frame()

taxo.amp$Species <- gsub("_sp.", "",as.character(taxo.amp$Species))

taxo.amp <- taxo.amp %>%
  select(c("Phylum", "Class", "Order", "Family", "Genus", "Species")) %>% 
  rownames_to_column("q_amplicon") %>% 
  mutate(Species = recode(Species,
                          "Burkholderia-Caballeronia-Paraburkholderia" = "BC-Paraburkholderia",
                          "Allorhizobium-Neorhizobium-Pararhizobium-Rhizobium" = "ANP-Rhizobium")) %>% 
  mutate(Genus = recode(Genus,
                        "Methylobacterium-Methylorubrum" = "Methylobacterium",
                        "Burkholderia-Caballeronia-Paraburkholderia" = "BC-Paraburkholderia",
                        "Allorhizobium-Neorhizobium-Pararhizobium-Rhizobium" = "ANP-Rhizobium")) %>% 
  mutate(Order = recode(Order,
                        "Enterobacterales" = "Enterobacteriales"))

# save(taxo.amp, file = "outputs/r/2_setup/taxo-amp.RData")

# taxo.pb <- as.data.frame(read_csv("outputs/r/2_setup/bh-tax-train-set-ncbi.csv"))

taxo.pb <- taxo.pb %>%
  select(c("db_pb", "Phylum", "Class", "Order", "Family", "Genus", "Species")) %>% 
  mutate(Species = recode(Species,
                        "Esherichica/Shigella" = "Escherichia-Shigella")) %>%
  mutate(Genus = recode(Genus,
                        "Esherichica/Shigella" = "Escherichia-Shigella")) %>%
  mutate(Family = recode(Family,
                        "Methylobacteriaceae" = "Beijerinckiaceae",
                        "Bradyrhizobiaceae" = "Xanthobacteraceae")) %>%
  mutate(Family = ifelse(Genus == "Brevibacillus" & Family == "Paenibacillaceae", "Brevibacillaceae", Family)) %>%
  mutate(Family = ifelse(Genus == "Pantoea" & Family == "Enterobacteriaceae", "Erwiniaceae", Family)) %>%
  mutate(Family = ifelse(Genus == "Erwinia" & Family == "Enterobacteriaceae", "Erwiniaceae", Family)) %>%
  mutate(Order = ifelse(Genus == "Brevibacillus" & Order == "Bacillales", "Brevibacillales", Order)) %>%
  mutate(Order = ifelse(Genus == "Paenibacillus" & Order == "Bacillales", "Paenibacillales", Order)) %>%
  mutate(Class = recode(Class,
                        "Betaproteobacteria" = "Gammaproteobacteria",
                        "Sphingobacteriia" = "Bacteroidia")) %>% 
  mutate(Phylum = recode(Phylum,
                        "Bacteroidetes" = "Bacteroidota",
                        "Actinobacteria" = "Actinobacteriota"))

# save(taxo.pb, file = "outputs/r/2_setup/taxo-pb.RData")

# Get assignment tables
c100 <- read.table(
  "outputs/galaxy/mb-cov100.txt",
  sep = "\t", quote = '', header = T, fill = T) #149314 assignments

check_pb <- c100 %>% select(db_pb) %>% unique() #419 ASVs
check_amp <- c100 %>% select(q_amplicon) %>% unique() #2513 ASVs

c100_strat <- c100 %>% 
  select(q_amplicon, id) %>% 
  mutate(type = 
           ifelse(id <= 85.93, "Uncultured", # "[77.35, 85.93]"
                       ifelse(id <= 87.95, "(85.93, 87.95]",
                              ifelse(id <= 90.1, "(87.95, 90.1]",
                                     ifelse(id <= 92.9, "(90.1, 92.9]",
                                            ifelse(id <= 96.55, "(92.9, 96.55]",
                                                   ifelse(id <= 99, "(96.55, 99]",
                                                         "(99, 100]"))))))) %>%
  mutate(type = factor(type, levels = c("(99, 100]", "(96.55, 99]", "(92.9, 96.55]", "(90.1, 92.9]", 
                                        "(87.95, 90.1]", "(85.93, 87.95]", "Uncultured"))) %>% # "[77.35, 85.93]"
  arrange(type) %>%
  distinct(q_amplicon, .keep_all = T) # 2513

check_unique <- c100_strat %>% select(q_amplicon) %>% unique() # 2513
check_type <- c100_strat  %>% group_by(type) %>% summarize(n = n_distinct(q_amplicon)); check_type
sum(check_type$n) # 2513

x <- setdiff(c100_strat$q_amplicon, rownames(otu_table(bh))) %>% as.data.frame()

c100_strat86 <- subset(c100_strat, c100_strat$type != "Uncultured")

# save(c100, file = "outputs/r/2_setup/c100.RData")
# save(c100_strat86, file = "outputs/r/2_setup/c100-strat86.RData")
```
